In [1]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms, models
from torch.utils.data import DataLoader
from torch.utils.data import random_split
import time
import os

In [3]:
def train_resnet(data_dir, num_classes, epochs=5, batch_size=8, lr=0.001, save_name='wytrenowany_resnet.pth'):

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    print(f"--- START ---")
    print(f"Device: {device} | Classes: {num_classes} | Epochs: {epochs}")

    if not os.path.exists(data_dir):
        print(f"BŁĄD: Folder {data_dir} nie istnieje!")
        return

    stat_transforms = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor()
    ])

    dataset = datasets.ImageFolder(root=data_dir, transform=stat_transforms)
    loader = DataLoader(dataset, batch_size=64, shuffle=False)

    mean = 0.
    std = 0.
    total_images = 0

    for images, _ in loader:
        batch_samples = images.size(0)
        images = images.view(batch_samples, images.size(1), -1)

        mean += images.mean(2).sum(0)
        std += images.std(2).sum(0)
        total_images += batch_samples

    mean /= total_images
    std /= total_images

    data_transforms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize(mean.tolist(), std.tolist())
    ])

    # 2. Dataset
    dataset = datasets.ImageFolder(root=data_dir, transform=data_transforms)
    dataset_size = len(dataset)

    # 80% train / 20% val
    train_size = int(0.8 * dataset_size)
    val_size = dataset_size - train_size

    train_dataset, val_dataset = random_split(dataset, [train_size, val_size])

    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)

    print(f"Train: {train_size} | Val: {val_size}")

    # Model
    model = models.resnet18(weights=None)
    num_ftrs = model.fc.in_features
    model.fc = nn.Linear(num_ftrs, num_classes)
    model = model.to(device)

    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=lr)
    best_acc = 0.0

    # Trening
    for epoch in range(epochs):
        start_time = time.time()

        # TRAIN
        model.train()
        running_loss = 0.0
        running_corrects = 0

        for inputs, labels in train_loader:
            inputs, labels = inputs.to(device), labels.to(device)

            optimizer.zero_grad()
            outputs = model(inputs)
            _, preds = torch.max(outputs, 1)
            loss = criterion(outputs, labels)

            loss.backward()
            optimizer.step()

            running_loss += loss.item() * inputs.size(0)
            running_corrects += torch.sum(preds == labels).item()

        epoch_loss = running_loss / train_size
        epoch_acc = running_corrects / train_size

        # VALIDATION
        model.eval()
        val_corrects = 0

        with torch.no_grad():
            for inputs, labels in val_loader:
                inputs, labels = inputs.to(device), labels.to(device)
                outputs = model(inputs)
                _, preds = torch.max(outputs, 1)
                val_corrects += torch.sum(preds == labels).item()

        val_acc = val_corrects / val_size

        print(f'Epoch {epoch+1}/{epochs} | '
              f'Train Acc: {epoch_acc:.4f} | '
              f'Val Acc: {val_acc:.4f} | '
              f'Time: {time.time()-start_time:.0f}s')

        if val_acc > best_acc:
            best_acc = val_acc
            torch.save(model.state_dict(), save_name)

    print(f"Best model saved as: {save_name}")
    return model

## puścić jeszcze raz

In [ ]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_niezalezne/FunnyBirds/train',
    num_classes = 4,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_niezalezne.pth'
)

--- START ---
Device: cpu | Classes: 4 | Epochs: 15
Train: 3200 | Val: 800
Epoch 1/15 | Train Acc: 0.4569 | Val Acc: 0.4925 | Time: 130s
Epoch 2/15 | Train Acc: 0.7334 | Val Acc: 0.7175 | Time: 128s
Epoch 3/15 | Train Acc: 0.9353 | Val Acc: 0.5363 | Time: 128s
Epoch 4/15 | Train Acc: 0.9731 | Val Acc: 0.7475 | Time: 129s
Epoch 5/15 | Train Acc: 0.9788 | Val Acc: 0.7688 | Time: 129s
Epoch 6/15 | Train Acc: 0.9862 | Val Acc: 0.6550 | Time: 129s
Epoch 7/15 | Train Acc: 0.9959 | Val Acc: 0.9513 | Time: 128s
Epoch 8/15 | Train Acc: 0.9969 | Val Acc: 0.8712 | Time: 128s
Epoch 9/15 | Train Acc: 0.9969 | Val Acc: 0.8788 | Time: 129s
Epoch 10/15 | Train Acc: 0.9888 | Val Acc: 0.6713 | Time: 130s
Epoch 11/15 | Train Acc: 0.9850 | Val Acc: 0.8175 | Time: 131s
Epoch 12/15 | Train Acc: 0.9953 | Val Acc: 0.8650 | Time: 129s
Epoch 13/15 | Train Acc: 0.9981 | Val Acc: 0.9300 | Time: 128s
Epoch 14/15 | Train Acc: 0.9988 | Val Acc: 0.9475 | Time: 128s
Epoch 15/15 | Train Acc: 0.9997 | Val Acc: 0.9313 | 

--- START --- lr 0.0001 batchsize 32
Device: cpu | Classes: 4 | Epochs: 15
Train: 3200 | Val: 800
Epoch 1/15 | Train Acc: 0.4569 | Val Acc: 0.4925 | Time: 130s
Epoch 2/15 | Train Acc: 0.7334 | Val Acc: 0.7175 | Time: 128s
Epoch 3/15 | Train Acc: 0.9353 | Val Acc: 0.5363 | Time: 128s
Epoch 4/15 | Train Acc: 0.9731 | Val Acc: 0.7475 | Time: 129s
Epoch 5/15 | Train Acc: 0.9788 | Val Acc: 0.7688 | Time: 129s
Epoch 6/15 | Train Acc: 0.9862 | Val Acc: 0.6550 | Time: 129s
Epoch 7/15 | Train Acc: 0.9959 | Val Acc: 0.9513 | Time: 128s
Epoch 8/15 | Train Acc: 0.9969 | Val Acc: 0.8712 | Time: 128s
Epoch 9/15 | Train Acc: 0.9969 | Val Acc: 0.8788 | Time: 129s
Epoch 10/15 | Train Acc: 0.9888 | Val Acc: 0.6713 | Time: 130s
Epoch 11/15 | Train Acc: 0.9850 | Val Acc: 0.8175 | Time: 131s
Epoch 12/15 | Train Acc: 0.9953 | Val Acc: 0.8650 | Time: 129s
Epoch 13/15 | Train Acc: 0.9981 | Val Acc: 0.9300 | Time: 128s
Epoch 14/15 | Train Acc: 0.9988 | Val Acc: 0.9475 | Time: 128s
Epoch 15/15 | Train Acc: 0.9997 | Val Acc: 0.9313 | Time: 129s
Best model saved as: model_hierarchia_nzal.pth

In [ ]:
model_eksperyment_2 = train_resnet(
    data_dir = './tests/testy_hierararchiczne_basic/FunnyBirds/train',
    num_classes = 3,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_hierarchiczne_basic.pth'
)

--- START ---
Device: cpu | Classes: 3 | Epochs: 15
Train: 2400 | Val: 600
Epoch 1/15 | Train Acc: 0.5846 | Val Acc: 0.6750 | Time: 106s
Epoch 2/15 | Train Acc: 0.7717 | Val Acc: 0.6783 | Time: 104s
Epoch 3/15 | Train Acc: 0.9033 | Val Acc: 0.5383 | Time: 102s
Epoch 4/15 | Train Acc: 0.9692 | Val Acc: 0.6783 | Time: 106s
Epoch 5/15 | Train Acc: 0.9946 | Val Acc: 0.7100 | Time: 104s
Epoch 6/15 | Train Acc: 0.9967 | Val Acc: 0.6067 | Time: 104s
Epoch 7/15 | Train Acc: 0.9971 | Val Acc: 0.8550 | Time: 99s
Epoch 8/15 | Train Acc: 0.9992 | Val Acc: 0.8633 | Time: 97s
Epoch 9/15 | Train Acc: 0.9971 | Val Acc: 0.7400 | Time: 97s
Epoch 10/15 | Train Acc: 0.9775 | Val Acc: 0.7767 | Time: 97s
Epoch 11/15 | Train Acc: 0.9788 | Val Acc: 0.8567 | Time: 97s
Epoch 12/15 | Train Acc: 0.9925 | Val Acc: 0.8350 | Time: 97s
Epoch 13/15 | Train Acc: 0.9996 | Val Acc: 0.8883 | Time: 97s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.8917 | Time: 97s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9083 | Time: 97

--- START --- lr 0.0001 batchsize 32
Device: cpu | Classes: 3 | Epochs: 15
Train: 2400 | Val: 600
Epoch 1/15 | Train Acc: 0.5846 | Val Acc: 0.6750 | Time: 106s
Epoch 2/15 | Train Acc: 0.7717 | Val Acc: 0.6783 | Time: 104s
Epoch 3/15 | Train Acc: 0.9033 | Val Acc: 0.5383 | Time: 102s
Epoch 4/15 | Train Acc: 0.9692 | Val Acc: 0.6783 | Time: 106s
Epoch 5/15 | Train Acc: 0.9946 | Val Acc: 0.7100 | Time: 104s
Epoch 6/15 | Train Acc: 0.9967 | Val Acc: 0.6067 | Time: 104s
Epoch 7/15 | Train Acc: 0.9971 | Val Acc: 0.8550 | Time: 99s
Epoch 8/15 | Train Acc: 0.9992 | Val Acc: 0.8633 | Time: 97s
Epoch 9/15 | Train Acc: 0.9971 | Val Acc: 0.7400 | Time: 97s
Epoch 10/15 | Train Acc: 0.9775 | Val Acc: 0.7767 | Time: 97s
Epoch 11/15 | Train Acc: 0.9788 | Val Acc: 0.8567 | Time: 97s
Epoch 12/15 | Train Acc: 0.9925 | Val Acc: 0.8350 | Time: 97s
Epoch 13/15 | Train Acc: 0.9996 | Val Acc: 0.8883 | Time: 97s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.8917 | Time: 97s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9083 | Time: 97s
Best model saved as: model_hierarchia_hier.pth

In [ ]:
model_eksperyment_4 = train_resnet(
    data_dir = './tests/testy_probabilistyczne/FunnyBirds/train',
    num_classes = 2,
    epochs = 10,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_testy_probabilistyczne.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 10
Train: 1600 | Val: 400
Epoch 1/10 | Train Acc: 0.6913 | Val Acc: 0.8550 | Time: 138s
Epoch 2/10 | Train Acc: 0.9587 | Val Acc: 0.5600 | Time: 137s
Epoch 3/10 | Train Acc: 0.9950 | Val Acc: 0.9375 | Time: 137s
Epoch 4/10 | Train Acc: 0.9988 | Val Acc: 0.9300 | Time: 127s
Epoch 5/10 | Train Acc: 1.0000 | Val Acc: 0.8550 | Time: 125s
Epoch 6/10 | Train Acc: 1.0000 | Val Acc: 0.9500 | Time: 126s
Epoch 7/10 | Train Acc: 1.0000 | Val Acc: 0.9500 | Time: 125s
Epoch 8/10 | Train Acc: 0.9994 | Val Acc: 0.9525 | Time: 127s
Epoch 9/10 | Train Acc: 1.0000 | Val Acc: 0.9425 | Time: 125s
Epoch 10/10 | Train Acc: 1.0000 | Val Acc: 0.9450 | Time: 125s
Best model saved as: model_testy_probabilistyczne.pth


In [8]:
model_eksperyment_5 = train_resnet(
    data_dir = './tests/testy_probabilistyczne_szum/FunnyBirds/train',
    num_classes = 2,
    epochs = 10,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_testy_probabilistyczne_szum.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 10
Train: 1600 | Val: 400
Epoch 1/10 | Train Acc: 0.7013 | Val Acc: 0.6625 | Time: 86s
Epoch 2/10 | Train Acc: 0.9456 | Val Acc: 0.8825 | Time: 65s
Epoch 3/10 | Train Acc: 0.9919 | Val Acc: 0.7975 | Time: 63s
Epoch 4/10 | Train Acc: 0.9975 | Val Acc: 0.8950 | Time: 62s
Epoch 5/10 | Train Acc: 0.9994 | Val Acc: 0.9550 | Time: 59s
Epoch 6/10 | Train Acc: 0.9994 | Val Acc: 0.9375 | Time: 63s
Epoch 7/10 | Train Acc: 1.0000 | Val Acc: 0.9475 | Time: 65s
Epoch 8/10 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 63s
Epoch 9/10 | Train Acc: 1.0000 | Val Acc: 0.9400 | Time: 64s
Epoch 10/10 | Train Acc: 1.0000 | Val Acc: 0.9425 | Time: 64s
Best model saved as: model_testy_probabilistyczne_szum.pth


In [9]:
model_eksperyment_6 = train_resnet(
    data_dir = './tests/testy_hierarchiczne/FunnyBirds/train',
    num_classes = 2,
    epochs = 10,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_testy_hierarchiczne.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 10
Train: 1600 | Val: 400
Epoch 1/10 | Train Acc: 0.6275 | Val Acc: 0.5375 | Time: 65s
Epoch 2/10 | Train Acc: 0.8781 | Val Acc: 0.6300 | Time: 66s
Epoch 3/10 | Train Acc: 0.9637 | Val Acc: 0.4725 | Time: 67s
Epoch 4/10 | Train Acc: 0.9938 | Val Acc: 0.5775 | Time: 68s
Epoch 5/10 | Train Acc: 0.9894 | Val Acc: 0.8225 | Time: 71s
Epoch 6/10 | Train Acc: 0.9831 | Val Acc: 0.8150 | Time: 69s
Epoch 7/10 | Train Acc: 0.9800 | Val Acc: 0.5600 | Time: 65s
Epoch 8/10 | Train Acc: 0.9825 | Val Acc: 0.8275 | Time: 65s
Epoch 9/10 | Train Acc: 0.9900 | Val Acc: 0.6450 | Time: 66s
Epoch 10/10 | Train Acc: 0.9931 | Val Acc: 0.8425 | Time: 66s
Best model saved as: model_testy_hierarchiczne.pth


In [6]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_4b/FunnyBirds/train',
    num_classes = 2,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_4b.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 15
Train: 1600 | Val: 400
Epoch 1/15 | Train Acc: 0.6494 | Val Acc: 0.5625 | Time: 63s
Epoch 2/15 | Train Acc: 0.9444 | Val Acc: 0.7050 | Time: 62s
Epoch 3/15 | Train Acc: 0.9856 | Val Acc: 0.8525 | Time: 62s
Epoch 4/15 | Train Acc: 0.9994 | Val Acc: 0.9475 | Time: 64s
Epoch 5/15 | Train Acc: 1.0000 | Val Acc: 0.9625 | Time: 63s
Epoch 6/15 | Train Acc: 1.0000 | Val Acc: 0.9650 | Time: 63s
Epoch 7/15 | Train Acc: 1.0000 | Val Acc: 0.9700 | Time: 63s
Epoch 8/15 | Train Acc: 1.0000 | Val Acc: 0.9625 | Time: 63s
Epoch 9/15 | Train Acc: 1.0000 | Val Acc: 0.9600 | Time: 63s
Epoch 10/15 | Train Acc: 1.0000 | Val Acc: 0.9675 | Time: 64s
Epoch 11/15 | Train Acc: 1.0000 | Val Acc: 0.9675 | Time: 64s
Epoch 12/15 | Train Acc: 1.0000 | Val Acc: 0.9725 | Time: 63s
Epoch 13/15 | Train Acc: 1.0000 | Val Acc: 0.9700 | Time: 63s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.9725 | Time: 66s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9700 | Time: 63s
Best

In [7]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_prob100_szum/FunnyBirds/train',
    num_classes = 2,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_prob100_szum.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 15
Train: 1600 | Val: 400
Epoch 1/15 | Train Acc: 0.6631 | Val Acc: 0.7100 | Time: 63s
Epoch 2/15 | Train Acc: 0.9431 | Val Acc: 0.9125 | Time: 64s
Epoch 3/15 | Train Acc: 0.9844 | Val Acc: 0.8925 | Time: 64s
Epoch 4/15 | Train Acc: 0.9956 | Val Acc: 0.9300 | Time: 64s
Epoch 5/15 | Train Acc: 0.9994 | Val Acc: 0.9325 | Time: 64s
Epoch 6/15 | Train Acc: 1.0000 | Val Acc: 0.9425 | Time: 63s
Epoch 7/15 | Train Acc: 1.0000 | Val Acc: 0.9600 | Time: 64s
Epoch 8/15 | Train Acc: 1.0000 | Val Acc: 0.9550 | Time: 63s
Epoch 9/15 | Train Acc: 1.0000 | Val Acc: 0.9500 | Time: 63s
Epoch 10/15 | Train Acc: 1.0000 | Val Acc: 0.9650 | Time: 63s
Epoch 11/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 63s
Epoch 12/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 63s
Epoch 13/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 70s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 64s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 64s
Best

In [8]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_prob60-40_szum/FunnyBirds/train',
    num_classes = 2,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_prob60-40_szum.pth'
)

--- START ---
Device: cpu | Classes: 2 | Epochs: 15
Train: 1600 | Val: 400
Epoch 1/15 | Train Acc: 0.6450 | Val Acc: 0.7950 | Time: 60s
Epoch 2/15 | Train Acc: 0.9325 | Val Acc: 0.8925 | Time: 58s
Epoch 3/15 | Train Acc: 0.9906 | Val Acc: 0.9075 | Time: 64s
Epoch 4/15 | Train Acc: 0.9969 | Val Acc: 0.9125 | Time: 69s
Epoch 5/15 | Train Acc: 0.9994 | Val Acc: 0.9150 | Time: 65s
Epoch 6/15 | Train Acc: 1.0000 | Val Acc: 0.8500 | Time: 64s
Epoch 7/15 | Train Acc: 1.0000 | Val Acc: 0.9225 | Time: 65s
Epoch 8/15 | Train Acc: 1.0000 | Val Acc: 0.9175 | Time: 64s
Epoch 9/15 | Train Acc: 1.0000 | Val Acc: 0.9325 | Time: 64s
Epoch 10/15 | Train Acc: 1.0000 | Val Acc: 0.9200 | Time: 64s
Epoch 11/15 | Train Acc: 1.0000 | Val Acc: 0.9500 | Time: 64s
Epoch 12/15 | Train Acc: 1.0000 | Val Acc: 0.9575 | Time: 64s
Epoch 13/15 | Train Acc: 1.0000 | Val Acc: 0.9550 | Time: 64s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.9550 | Time: 64s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9525 | Time: 64s
Best

In [4]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_basic_hierarchia100/FunnyBirds/train',
    num_classes = 3,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_basic_hierarchia100.pth'
)

--- START ---
Device: cpu | Classes: 3 | Epochs: 15
Train: 2400 | Val: 600
Epoch 1/15 | Train Acc: 0.8125 | Val Acc: 0.9000 | Time: 88s
Epoch 2/15 | Train Acc: 0.9883 | Val Acc: 0.8633 | Time: 89s
Epoch 3/15 | Train Acc: 0.9962 | Val Acc: 0.9767 | Time: 95s
Epoch 4/15 | Train Acc: 0.9996 | Val Acc: 0.9883 | Time: 94s
Epoch 5/15 | Train Acc: 0.9992 | Val Acc: 0.9883 | Time: 93s
Epoch 6/15 | Train Acc: 0.9967 | Val Acc: 0.9450 | Time: 94s
Epoch 7/15 | Train Acc: 0.9971 | Val Acc: 0.9883 | Time: 97s
Epoch 8/15 | Train Acc: 0.9988 | Val Acc: 0.8983 | Time: 96s
Epoch 9/15 | Train Acc: 1.0000 | Val Acc: 0.9850 | Time: 96s
Epoch 10/15 | Train Acc: 0.9996 | Val Acc: 0.9867 | Time: 96s
Epoch 11/15 | Train Acc: 0.9996 | Val Acc: 0.9883 | Time: 96s
Epoch 12/15 | Train Acc: 1.0000 | Val Acc: 0.9933 | Time: 95s
Epoch 13/15 | Train Acc: 1.0000 | Val Acc: 0.9917 | Time: 95s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.9900 | Time: 96s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.9900 | Time: 96s
Best

In [5]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_basic_hierarchia80/FunnyBirds/train',
    num_classes = 3,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_basic_hierarchia80.pth'
)

--- START ---
Device: cpu | Classes: 3 | Epochs: 15
Train: 2400 | Val: 600
Epoch 1/15 | Train Acc: 0.7033 | Val Acc: 0.5817 | Time: 94s
Epoch 2/15 | Train Acc: 0.8875 | Val Acc: 0.7433 | Time: 94s
Epoch 3/15 | Train Acc: 0.9550 | Val Acc: 0.7800 | Time: 98s
Epoch 4/15 | Train Acc: 0.9754 | Val Acc: 0.8133 | Time: 96s
Epoch 5/15 | Train Acc: 0.9912 | Val Acc: 0.8433 | Time: 94s
Epoch 6/15 | Train Acc: 0.9929 | Val Acc: 0.8150 | Time: 95s
Epoch 7/15 | Train Acc: 0.9938 | Val Acc: 0.7117 | Time: 94s
Epoch 8/15 | Train Acc: 0.9962 | Val Acc: 0.7467 | Time: 94s
Epoch 9/15 | Train Acc: 0.9983 | Val Acc: 0.8400 | Time: 95s
Epoch 10/15 | Train Acc: 0.9933 | Val Acc: 0.8533 | Time: 95s
Epoch 11/15 | Train Acc: 0.9888 | Val Acc: 0.7183 | Time: 95s
Epoch 12/15 | Train Acc: 0.9858 | Val Acc: 0.8583 | Time: 94s
Epoch 13/15 | Train Acc: 0.9933 | Val Acc: 0.8817 | Time: 93s
Epoch 14/15 | Train Acc: 0.9992 | Val Acc: 0.8750 | Time: 94s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.8883 | Time: 94s
Best

In [7]:
model_eksperyment_1 = train_resnet(
    data_dir = './tests/testy_basic_hierarchia60/FunnyBirds/train',
    num_classes = 3,
    epochs = 15,
    batch_size= 32,
    lr = 0.0001,
    save_name = 'model_basic_hierarchia60.pth'
)

--- START ---
Device: cpu | Classes: 3 | Epochs: 15
Train: 2400 | Val: 600
Epoch 1/15 | Train Acc: 0.6167 | Val Acc: 0.6783 | Time: 87s
Epoch 2/15 | Train Acc: 0.8438 | Val Acc: 0.7050 | Time: 91s
Epoch 3/15 | Train Acc: 0.9371 | Val Acc: 0.6883 | Time: 93s
Epoch 4/15 | Train Acc: 0.9742 | Val Acc: 0.7950 | Time: 99s
Epoch 5/15 | Train Acc: 0.9883 | Val Acc: 0.8333 | Time: 99s
Epoch 6/15 | Train Acc: 0.9942 | Val Acc: 0.7550 | Time: 99s
Epoch 7/15 | Train Acc: 0.9992 | Val Acc: 0.7967 | Time: 99s
Epoch 8/15 | Train Acc: 1.0000 | Val Acc: 0.8700 | Time: 99s
Epoch 9/15 | Train Acc: 1.0000 | Val Acc: 0.8650 | Time: 98s
Epoch 10/15 | Train Acc: 1.0000 | Val Acc: 0.8717 | Time: 99s
Epoch 11/15 | Train Acc: 1.0000 | Val Acc: 0.8650 | Time: 99s
Epoch 12/15 | Train Acc: 1.0000 | Val Acc: 0.8600 | Time: 99s
Epoch 13/15 | Train Acc: 1.0000 | Val Acc: 0.8733 | Time: 99s
Epoch 14/15 | Train Acc: 1.0000 | Val Acc: 0.8767 | Time: 99s
Epoch 15/15 | Train Acc: 1.0000 | Val Acc: 0.8833 | Time: 100s
Bes

In [2]:
def extract_features_and_weights(data_dir, model_path, model_weights, model_features, num_classes, batch_size=64):
    """
    Ekstrahuje wektory cech (features) oraz wagi klasyfikatora (weights) 
    z wytrenowanego modelu ResNet18 i zapisuje je pod wskazanymi nazwami.
    """
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

    stat_transforms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor()
    ])
    temp_dataset = datasets.ImageFolder(root=data_dir, transform=stat_transforms)
    temp_loader = DataLoader(temp_dataset, batch_size=batch_size, shuffle=False)

    mean, std, total_images = 0., 0., 0
    for images, _ in temp_loader:
        batch_samples = images.size(0)
        images = images.view(batch_samples, images.size(1), -1)
        mean += images.mean(2).sum(0)
        std += images.std(2).sum(0)
        total_images += batch_samples

    mean /= total_images
    std /= total_images

    data_transforms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize(mean.tolist(), std.tolist())
    ])
    dataset = datasets.ImageFolder(root=data_dir, transform=data_transforms)
    loader = DataLoader(dataset, batch_size=batch_size, shuffle=False) 

    # 3. Wczytanie modelu
    model = models.resnet18(weights=None)
    num_ftrs = model.fc.in_features
    model.fc = nn.Linear(num_ftrs, num_classes)
    
    model.load_state_dict(torch.load(model_path, map_location=device))
    model = model.to(device)
    model.eval()

    weights_to_save = model.fc.weight.detach().cpu().clone()
    torch.save(weights_to_save, model_weights)
    print(f"-> Zapisano wagi: {model_weights}")

    model.fc = nn.Identity()

    all_features = []
    all_labels = []

    with torch.no_grad():
        for inputs, labels in loader:
            inputs = inputs.to(device)
            features = model(inputs) 
            all_features.append(features.cpu())
            all_labels.append(labels.cpu())

    all_features = torch.cat(all_features)
    all_labels = torch.cat(all_labels)

    features_dict = {
        'features': all_features,
        'labels': all_labels
    }
    torch.save(features_dict, model_features)
    
    print(f"-> Zapisano wektory: {model_features}")
    print(f"Kształt macierzy cech: {all_features.shape}")
    print("--- ZAKOŃCZONO ---")

In [ ]:
extract_features_and_weights(
    data_dir='./testy_niezalezne/FunnyBirds/train',
    model_path='model_hierarchia_nzal.pth', 
    model_weights='model_niezalezne_weights.pt',
    model_features='model_niezalezne_features.pt',
    num_classes=4
)


--- ROZPOCZYNAM EKSTRAKCJĘ: ./testy_nzal/FunnyBirds/train ---
-> Zapisano wagi: model_nzal_weights.pt
-> Zapisano wektory: model_nzal_features.pt
Kształt macierzy cech: torch.Size([4000, 512])
--- ZAKOŃCZONO ---


In [10]:
extract_features_and_weights(
    data_dir='./testy_hier/FunnyBirds/train',
    model_path='model_hierarchia_hier.pth', 
    model_weights='model_hier_weights.pt',
    model_features='model_hier_features.pt',
    num_classes=3
)


--- ROZPOCZYNAM EKSTRAKCJĘ: ./testy_hier/FunnyBirds/train ---
-> Zapisano wagi: model_hier_weights.pt
-> Zapisano wektory: model_hier_features.pt
Kształt macierzy cech: torch.Size([3000, 512])
--- ZAKOŃCZONO ---


In [7]:
extract_features_and_weights(
    data_dir='tests/testy_hierarchiczne/FunnyBirds/train',
    model_path='models/full/model_testy_hierarchiczne.pth', 
    model_weights='models/weights/model_hierarchiczne_weights.pt',
    model_features='models/features/model_hierarchiczne_features.pt',
    num_classes=2
)


--- ROZPOCZYNAM EKSTRAKCJĘ: tests/testy_hierarchiczne/FunnyBirds/train ---
-> Zapisano wagi: models/weights/model_hierarchiczne_weights.pt
-> Zapisano wektory: models/features/model_hierarchiczne_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [9]:
extract_features_and_weights(
    data_dir='tests/testy_probabilistyczne/FunnyBirds/train',
    model_path='models/full/model_testy_probabilistyczne.pth', 
    model_weights='models/weights/model_probabilistyczne_weights.pt',
    model_features='models/features/model_probabilistyczne_features.pt',
    num_classes=2
)


--- ROZPOCZYNAM EKSTRAKCJĘ: tests/testy_probabilistyczne/FunnyBirds/train ---
-> Zapisano wagi: models/weights/model_probabilistyczne_weights.pt
-> Zapisano wektory: models/features/model_probabilistyczne_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [10]:
extract_features_and_weights(
    data_dir='tests/testy_probabilistyczne_szum/FunnyBirds/train',
    model_path='models/full/model_testy_probabilistyczne_szum.pth', 
    model_weights='models/weights/model_probabilistyczne_szum_weights.pt',
    model_features='models/features/model_probabilistyczne_szum_features.pt',
    num_classes=2
)


--- ROZPOCZYNAM EKSTRAKCJĘ: tests/testy_probabilistyczne_szum/FunnyBirds/train ---
-> Zapisano wagi: models/weights/model_probabilistyczne_szum_weights.pt
-> Zapisano wektory: models/features/model_probabilistyczne_szum_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [5]:
extract_features_and_weights(
    data_dir='tests/testy_4b/FunnyBirds/train',
    model_path='model_4b.pth', 
    model_weights='model_4b_weights.pt',
    model_features='model_4b_features.pt',
    num_classes=2
)

-> Zapisano wagi: model_4b_weights.pt
-> Zapisano wektory: model_4b_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [6]:
extract_features_and_weights(
    data_dir='tests/testy_prob100_szum/FunnyBirds/train',
    model_path='model_prob100_szum.pth', 
    model_weights='model_prob100_szum_weights.pt',
    model_features='model_prob100_szum_features.pt',
    num_classes=2
)

-> Zapisano wagi: model_prob100_szum_weights.pt
-> Zapisano wektory: model_prob100_szum_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [8]:
extract_features_and_weights(
    data_dir='tests/testy_prob60-40_szum/FunnyBirds/train',
    model_path='model_prob60-40_szum.pth', 
    model_weights='model_prob60-40_szum_weights.pt',
    model_features='model_prob60-40_szum_features.pt',
    num_classes=2
)

-> Zapisano wagi: model_prob60-40_szum_weights.pt
-> Zapisano wektory: model_prob60-40_szum_features.pt
Kształt macierzy cech: torch.Size([2000, 512])
--- ZAKOŃCZONO ---


In [3]:
extract_features_and_weights(
    data_dir='tests/testy_basic_hierarchia60/FunnyBirds/train',
    model_path='models/full/model_basic_hierarchia60.pth', 
    model_weights='models/weights/model_basic_hierarchia60_weights.pt',
    model_features='models/features/model_basic_hierarchia60_features.pt',
    num_classes=3
)

-> Zapisano wagi: models/weights/model_basic_hierarchia60_weights.pt
-> Zapisano wektory: models/features/model_basic_hierarchia60_features.pt
Kształt macierzy cech: torch.Size([3000, 512])
--- ZAKOŃCZONO ---


In [4]:
extract_features_and_weights(
    data_dir='tests/testy_basic_hierarchia80/FunnyBirds/train',
    model_path='models/full/model_basic_hierarchia80.pth', 
    model_weights='models/weights/model_basic_hierarchia80_weights.pt',
    model_features='models/features/model_basic_hierarchia80_features.pt',
    num_classes=3
)

-> Zapisano wagi: models/weights/model_basic_hierarchia80_weights.pt
-> Zapisano wektory: models/features/model_basic_hierarchia80_features.pt
Kształt macierzy cech: torch.Size([3000, 512])
--- ZAKOŃCZONO ---


In [5]:
extract_features_and_weights(
    data_dir='tests/testy_basic_hierarchia100/FunnyBirds/train',
    model_path='models/full/model_basic_hierarchia100.pth', 
    model_weights='models/weights/model_basic_hierarchia100_weights.pt',
    model_features='models/features/model_basic_hierarchia100_features.pt',
    num_classes=3
)

-> Zapisano wagi: models/weights/model_basic_hierarchia100_weights.pt
-> Zapisano wektory: models/features/model_basic_hierarchia100_features.pt
Kształt macierzy cech: torch.Size([3000, 512])
--- ZAKOŃCZONO ---


In [3]:

def extract_features_all_layers(data_dir, model_path, prefix_save_name, num_classes, batch_size=64):
    """
    Ekstrahuje wektory cech z warstw: layer1, layer2, layer3, layer4.
    Używa mechanizmu Forward Hooks i Global Average Pooling.
    Zapisuje 4 osobne pliki .pt (np. 'moj_model_features_layer1.pt').
    """
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    print(f"--- START EKSTRAKCJI WIELOWARSTWOWEJ ---")

    # 1. Obliczanie statystyk (Średnia i Odchylenie Standardowe)
    stat_transforms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor()
    ])
    temp_dataset = datasets.ImageFolder(root=data_dir, transform=stat_transforms)
    temp_loader = DataLoader(temp_dataset, batch_size=batch_size, shuffle=False)

    mean, std, total_images = 0., 0., 0
    for images, _ in temp_loader:
        batch_samples = images.size(0)
        images = images.view(batch_samples, images.size(1), -1)
        mean += images.mean(2).sum(0)
        std += images.std(2).sum(0)
        total_images += batch_samples

    mean /= total_images
    std /= total_images

    # 2. Przygotowanie docelowego Dataloadera
    data_transforms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize(mean.tolist(), std.tolist())
    ])
    dataset = datasets.ImageFolder(root=data_dir, transform=data_transforms)
    loader = DataLoader(dataset, batch_size=batch_size, shuffle=False) 

    # 3. Wczytanie modelu
    model = models.resnet18(weights=None)
    num_ftrs = model.fc.in_features
    model.fc = nn.Linear(num_ftrs, num_classes)
    
    model.load_state_dict(torch.load(model_path, map_location=device))
    model = model.to(device)
    model.eval()

    # 4. MECHANIZM HACZYKÓW (FORWARD HOOKS)
    extracted_features = {
        'layer1': [], # Tutaj wpadną cechy (wymiar: 64)
        'layer2': [], # Tutaj wpadną cechy (wymiar: 128)
        'layer3': [], # Tutaj wpadną cechy (wymiar: 256)
        'layer4': []  # Tutaj wpadną cechy (wymiar: 512)
    }
    
    # Narzędzie do uśredniania obrazka do postaci płaskiego wektora (Global Average Pooling)
    pool = nn.AdaptiveAvgPool2d((1, 1))

    # Funkcja, która doczepia się do warstwy i podsłuchuje wyjście
    def get_hook(layer_name):
        def hook(module, input, output):
            # output ma kształt (Batch, Kanały, Wysokość, Szerokość)
            # pool zgniata (W, S) do punktu, a flatten przerabia to na płaski wektor
            pooled_output = pool(output).flatten(1)
            extracted_features[layer_name].append(pooled_output.cpu().detach())
        return hook

    # Zakładamy podsłuchy na główne bloki ResNetu
    hook_l1 = model.layer1.register_forward_hook(get_hook('layer1'))
    hook_l2 = model.layer2.register_forward_hook(get_hook('layer2'))
    hook_l3 = model.layer3.register_forward_hook(get_hook('layer3'))
    hook_l4 = model.layer4.register_forward_hook(get_hook('layer4'))

    all_labels = []

    # 5. Przepuszczenie obrazków przez model (Haczyki same zbiorą dane!)
    print("Przepuszczanie danych przez model...")
    with torch.no_grad():
        for inputs, labels in loader:
            inputs = inputs.to(device)
            _ = model(inputs) # Zwykły forward pass, tu dzieje się magia!
            all_labels.append(labels.cpu())

    # Zdejmujemy haczyki po robocie (dobra praktyka)
    hook_l1.remove()
    hook_l2.remove()
    hook_l3.remove()
    hook_l4.remove()

    all_labels = torch.cat(all_labels)

    # 6. Zapisywanie osobnych plików dla każdej warstwy
    for layer_name, features_list in extracted_features.items():
        layer_features = torch.cat(features_list)
        
        features_dict = {
            'features': layer_features,
            'labels': all_labels
        }
        
        save_path = f"{prefix_save_name}_{layer_name}.pt"
        torch.save(features_dict, save_path)
        
        print(f"-> Zapisano wektory {layer_name}: {save_path}")
        print(f"   Kształt macierzy: {layer_features.shape}")

    print("--- ZAKOŃCZONO EKSTRAKCJĘ ---")


In [4]:
experiments = [
    # ZBIÓR 1
    {"title": "Zbiór 1: Niezależność", "pt": "models/features/model_niezalezne_features.pt", "json": "tests/testy_niezalezne/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot02.glb"]},
    
    # ZBIÓR 2a
    {"title": "Zbiór 2a: Hierarchia Basic", "pt": "models/features/model_hierarchiczne_basic_features.pt", "json": "tests/testy_hierarchiczne_basic/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot02.glb"]},
    
    # ZBIÓR 2b (Hierarchiczne + Kolor)
    # Wariant 100%
    {"title": "2b (100%): Dziób vs Nogi", "pt": "models/features/model_basic_hierarchia100_features.pt", "json": "tests/testy_basic_hierarchia100/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot02.glb"]},
    {"title": "2b (100%): Dziób vs Skrzydła", "pt": "models/features/model_basic_hierarchia100_features.pt", "json": "tests/testy_basic_hierarchia100/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "wing_color", "valB": ["red", "green"]},
    {"title": "2b (100%): Nogi vs Skrzydła", "pt": "models/features/model_basic_hierarchia100_features.pt", "json": "tests/testy_basic_hierarchia100/FunnyBirds/dataset_train.json", "featA": "foot_model", "valA": ["foot01.glb", "foot02.glb"], "featB": "wing_color", "valB": ["green", "blue"]},

    # Wariant 80%
    {"title": "2b (80%): Dziób vs Nogi", "pt": "models/features/model_basic_hierarchia80_features.pt", "json": "tests/testy_basic_hierarchia80/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot02.glb"]},
    {"title": "2b (80%): Dziób vs Skrzydła", "pt": "models/features/model_basic_hierarchia80_features.pt", "json": "tests/testy_basic_hierarchia80/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "wing_color", "valB": ["red", "green"]},
    {"title": "2b (80%): Nogi vs Skrzydła", "pt": "models/features/model_basic_hierarchia80_features.pt", "json": "tests/testy_basic_hierarchia80/FunnyBirds/dataset_train.json", "featA": "foot_model", "valA": ["foot01.glb", "foot02.glb"], "featB": "wing_color", "valB": ["green", "blue"]},

    # Wariant 60%
    {"title": "2b (60%): Dziób vs Nogi", "pt": "models/features/model_basic_hierarchia60_features.pt", "json": "tests/testy_basic_hierarchia60/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot02.glb"]},
    {"title": "2b (60%): Dziób vs Skrzydła", "pt": "models/features/model_basic_hierarchia60_features.pt", "json": "tests/testy_basic_hierarchia60/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "wing_color", "valB": ["red", "green"]},
    {"title": "2b (60%): Nogi vs Skrzydła", "pt": "models/features/model_basic_hierarchia60_features.pt", "json": "tests/testy_basic_hierarchia60/FunnyBirds/dataset_train.json", "featA": "foot_model", "valA": ["foot01.glb", "foot02.glb"], "featB": "wing_color", "valB": ["green", "blue"]},

    # ZBIÓR 3a (Probabilistyczne)
    {"title": "Zbiór 3a: Probabilistyczne", "pt": "models/features/model_probabilistyczne_features.pt", "json": "tests/testy_probabilistyczne/FunnyBirds/dataset_train.json", "featA": "beak_model", "valA": ["beak01.glb", "beak02.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot04.glb"]},
    
    # ZBIÓR 3b, 4a, 4b, 4c (Oczy vs Nogi)
    {"title": "Zbiór 3b: 4b", "pt": "models/features/model_4b_features.pt", "json": "tests/testy_4b/FunnyBirds/dataset_train.json", "featA": "eye_model", "valA": ["eye01.glb", "eye03.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot03.glb"]},
    {"title": "Zbiór 4a: Prob100 szum", "pt": "models/features/model_prob100_szum_features.pt", "json": "tests/testy_prob100_szum/FunnyBirds/dataset_train.json", "featA": "eye_model", "valA": ["eye01.glb", "eye03.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot03.glb"]},
    {"title": "Zbiór 4b: Probabilistyczne szum", "pt": "models/features/model_probabilistyczne_szum_features.pt", "json": "tests/testy_probabilistyczne_szum/FunnyBirds/dataset_train.json", "featA": "eye_model", "valA": ["eye01.glb", "eye03.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot03.glb"]},
    {"title": "Zbiór 4c: Prob 60-40 szum", "pt": "models/features/model_prob60-40_szum_features.pt", "json": "tests/testy_prob60-40_szum/FunnyBirds/dataset_train.json", "featA": "eye_model", "valA": ["eye01.glb", "eye03.glb"], "featB": "foot_model", "valB": ["foot01.glb", "foot03.glb"]},
]


In [9]:

print("Analizowanie listy eksperymentów w poszukiwaniu unikalnych modeli...")

processed_models = set()

for exp in experiments:
    pt_path = exp["pt"]
    
    if pt_path in processed_models:
        continue
        
    processed_models.add(pt_path)
    
    print(f"\n{'='*40}")
    print(f"PRZYGOTOWANIE DO EKSTRAKCJI: {exp['title']}")
    print(f"{'='*40}")
    
    data_dir = os.path.dirname(exp["json"])
    
    normalized_path = pt_path.replace('\\', '/')
    filename = normalized_path.split('/')[-1] 
    
    model_filename = filename.replace('_features.pt', '.pth')
    model_path = os.path.join('models', 'full', model_filename)
    
    if not os.path.exists(model_path):
        alt_filename = model_filename.replace('model_', 'model_testy_')
        alt_model_path = os.path.join('models', 'full', alt_filename)
        if os.path.exists(alt_model_path):
            model_path = alt_model_path
            
    prefix_save_name = pt_path.replace('.pt', '')
    
    if os.path.exists(f"{prefix_save_name}_layer4.pt"):
        print(f"-> ZNALEZIONO GOTOWE PLIKI dla {exp['title']}!")
        print(f"-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).")
        continue
    
    if "Zbiór 1" in exp["title"]:
        num_classes = 4
    elif "2a" in exp["title"] or "2b" in exp["title"]:
        num_classes = 3
    else:
        num_classes = 2
        
    print(f"Folder zdjęć: {data_dir}")
    print(f"Wagi modelu : {model_path}")
    print(f"Liczba klas : {num_classes}")
    print(f"Zapis do    : {prefix_save_name}_layerX.pt")
    
    if not os.path.exists(model_path):
        print(f"UWAGA: Nie znaleziono pełnego modelu: {model_path}! Pomijam ten model.")
        continue
        
    try:
        extract_features_all_layers(
            data_dir=data_dir,
            model_path=model_path,
            prefix_save_name=prefix_save_name,
            num_classes=num_classes,
            batch_size=64
        )
    except Exception as e:
        print(f"BŁĄD podczas ekstrakcji dla {exp['title']}: {e}")

print("\nWSZYSTKIE UNIKALNE MODELE ZOSTAŁY PRZETWORZONE!")

Analizowanie listy eksperymentów w poszukiwaniu unikalnych modeli...

PRZYGOTOWANIE DO EKSTRAKCJI: Zbiór 1: Niezależność
-> ZNALEZIONO GOTOWE PLIKI dla Zbiór 1: Niezależność!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTOWANIE DO EKSTRAKCJI: Zbiór 2a: Hierarchia Basic
-> ZNALEZIONO GOTOWE PLIKI dla Zbiór 2a: Hierarchia Basic!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTOWANIE DO EKSTRAKCJI: 2b (100%): Dziób vs Nogi
-> ZNALEZIONO GOTOWE PLIKI dla 2b (100%): Dziób vs Nogi!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTOWANIE DO EKSTRAKCJI: 2b (80%): Dziób vs Nogi
-> ZNALEZIONO GOTOWE PLIKI dla 2b (80%): Dziób vs Nogi!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTOWANIE DO EKSTRAKCJI: 2b (60%): Dziób vs Nogi
-> ZNALEZIONO GOTOWE PLIKI dla 2b (60%): Dziób vs Nogi!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTOWANIE DO EKSTRAKCJI: Zbiór 3a: Probabilistyczne
-> ZNALEZIONO GOTOWE PLIKI dla Zbiór 3a: Probabilistyczne!
-> POMIJAM EKSTRAKCJĘ (Oszczędzamy czas).

PRZYGOTO